### PR TITLE
Waiting for a ready msg from server

### PR DESCRIPTION
**Pull request by trycatcher**
Server now sends a "Ready frame" when the on-recive Callback hooks up after which messages can safely be sent by the Client.This addresses http-kit/http-kit#318

### CODE REVIEW

File: src/thunks.js [hunk 1]
```diff
@@ -42,32 +42,41 @@ export function makeWSConnection() {
...
+
+      if (message.type === 'ready') {
+          wsConnection.send(JSON.stringify({
+          command: 'authenticate',
+          token: authToken
+        })); 
       }
       else {
-        dispatch(wsHandshakeFailed());
+        if (message['auth-status'] === 'success') {
```
Comment (jysandy) [inline src/thunks.js]: @trycatcher here if the first message is a `success` message and not a ready frame, the connection will still be considered authenticated and ready, even though the client never sent an authentication message.

Although this 'works', I think our implementation ought to be more correct. The client should only wait for the authentication success message after it has received the `ready` message and sent the `authenticate` message. Here's a state diagram to clarify:

waiting for ready frame --> waiting for authentication success --> receiving data messages

**Review by jysandy — CHANGES_REQUESTED** [review]

File: src/thunks.js (hunk 1 above)
Comment (trycatcher) [reply src/thunks.js]: @jysandy Changed in line with the state diagram. Additionally, if the first message is not a "Ready" the connection is closed. https://github.com/nilenso/time-tracker-web/blob/11a09e62beb247c88ce745964ddc55fadbdf5c92/src/thunks.js

**Comment by trycatcher** [comment]
@jysandy Pushed the change for this https://github.com/nilenso/time-tracker-web/pull/15/commits/11a09e62beb247c88ce745964ddc55fadbdf5c92

File: src/thunks.js [hunk 2]
```diff
@@ -42,32 +42,47 @@ export function makeWSConnection() {
...
+          }
+          else {
+            dispatch(wsHandshakeFailed());
+          }
         }
       }
       else {
-        dispatch(wsHandshakeFailed());
-      }
+        wsConnection.close();
```
Comment (jysandy) [inline src/thunks.js]: @trycatcher you should `dispatch(wsHandshakeFailed())` here, just like on line 79.

**Review by jysandy — CHANGES_REQUESTED** [review]

**Comment by jysandy** [comment]
@trycatcher once you make that super small change I'm actually OK with this being merged, so go ahead and merge this and https://github.com/nilenso/time-tracker/pull/82 after you make the change.

**Comment by jysandy** [comment]
As a bit of an aside: Redefining the callback is easy to understand if the connection doesn't have too many states. Another way of implementing this would have been to add a state variable, which can maybe take values like `WAITING_FOR_READY_MESSAGE` or `WAITING_FOR_AUTH_RESPONSE`, etc, and then in the callback just `switch` on that variable and do the appropriate thing. Fairly straightforward way of implementing a state machine.

**Comment by trycatcher** [comment]
@jysandy Added the dispatch. Will go ahead merging this and https://github.com/nilenso/time-tracker/pull/82

**Comment by trycatcher** [comment]
@jysandy I'm in favor of making the state transition explicit over redefining callbacks. I'll add a low-priority tech card so that we can come back and do it.

**Pull request merged**